crates/vm/levm/src/vm.rs

```diff
@@ -401,407 +390,24 @@ impl VM {
-            )?;
-        }
-
-        // 2. Return unused gas + gas refunds to the sender.
-        let max_gas = self.env.gas_limit;
-        let consumed_gas = report.gas_used;
-        let refunded_gas = report.gas_refunded.min(
-            consumed_gas
-                .checked_div(5)
-                .ok_or(VMError::Internal(InternalError::UndefinedState(-1)))?,
-        );
-        // "The max refundable proportion of gas was reduced from one half to one fifth by EIP-3529 by Buterin and Swende [2021] in the London release"
-        report.gas_refunded = refunded_gas;
-
-        let gas_to_return = max_gas
-            .checked_sub(consumed_gas)
-            .and_then(|gas| gas.checked_add(refunded_gas))
-            .ok_or(VMError::Internal(InternalError::UndefinedState(0)))?;
-
-        let wei_return_amount = self
-            .env
-            .gas_price
-            .checked_mul(U256::from(gas_to_return))
-            .ok_or(VMError::Internal(InternalError::UndefinedState(1)))?;
-
-        increase_account_balance(
-            &mut self.cache,
-            &mut self.db,
-            sender_address,
-            wei_return_amount,
-        )?;
-
-        // 3. Pay coinbase fee
-        let coinbase_address = self.env.coinbase;
-
-        let gas_to_pay_coinbase = consumed_gas
-            .checked_sub(refunded_gas)
-            .ok_or(VMError::Internal(InternalError::UndefinedState(2)))?;
-
-        let priority_fee_per_gas = self
-            .env
-            .gas_price
-            .checked_sub(self.env.base_fee_per_gas)
-            .ok_or(VMError::GasPriceIsLowerThanBaseFee)?;
-        let coinbase_fee = U256::from(gas_to_pay_coinbase)
-            .checked_mul(priority_fee_per_gas)
-            .ok_or(VMError::BalanceOverflow)?;
-
-        if coinbase_fee != U256::zero() {
-            increase_account_balance(
-                &mut self.cache,
-                &mut self.db,
-                coinbase_address,
-                coinbase_fee,
-            )?;
-        };
-
-        // 4. Destruct addresses in selfdestruct set.
-        // In Cancun the only addresses destroyed are contracts created in this transaction
-        let selfdestruct_set = self.accrued_substate.selfdestruct_set.clone();
-        for address in selfdestruct_set {
-            let account_to_remove = get_account_mut_vm(&mut self.cache, &self.db, address)?;
-            *account_to_remove = Account::default();
-        }
-
-        Ok(())
-    }
-
     pub fn execute(&mut self) -> Result<ExecutionReport, VMError> {
         let mut initial_call_frame = self
             .call_frames
             .pop()
             .ok_or(VMError::Internal(InternalError::CouldNotPopCallframe))?;
 
-        self.prepare_execution(&mut initial_call_frame)?;
+        // NOTE: ATTOW the default hook is created in VM::new(), so
+        // (in theory) _at least_ the default prepare execution should
+        // run
+        for hook in self.hooks.clone() {
```

Comment:
I believe this should be enough:
```suggestion
        for hook in self.hooks.iter() {
```

Reply:
Hmm, I'm getting the following error:

```
402 | ...or hook in self.hooks.iter() {
    |               -----------------
    |               |
    |               immutable borrow occurs here
    |               immutable borrow later used here
403 | ...   hook.prepare_execution(self, &mut initial_call_frame...
    |       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ mutable borrow occurs here
```